test/others/crit/test.sh

```diff
@@ -17,7 +17,8 @@ function _exit {
 
 function gen_imgs {
 	setsid ./loop.sh < /dev/null &> /dev/null &
-	PID=$!
+	sleep .1  # wait for loop.sh to start
```

Comment:
I am not a big fan of using sleeps to fix races ;). I think we can use pipes here:
https://github.com/opencontainers/runc/blob/master/tests/integration/checkpoint.bats#L24

Reply:
Hi @avagin, I was able to replace `sleep` with nested command substitution.
```
2021-06-06T11:12:02.5890420Z ++ ++ setsid ./loop.sh
2021-06-06T11:12:02.5894022Z jobs -r -p
2021-06-06T11:12:02.5900376Z + PID=658293
2021-06-06T11:12:02.5901311Z + /criu/criu/criu dump -v4 -o dump.log -D ./ -t 658293
```

Reply:
pls explain in more detail. I don't understand...

Reply:
The problem that was causing the test to fail was that `PID=!$` was executed before `setsid ./loop.sh`. These commands were not executed in the correct order on Alpine and Fedora Rawhide, as show below.
```bash
setsid ./loop.sh </dev/null &> /dev/null &
PID=$!
$CRIU dump -v4 -o dump.log -D ./ -t $PID
```
```
2021-05-30T13:36:43.8309467Z + PID=494959
2021-05-30T13:36:43.8310637Z + /criu/criu/criu dump -v4 -o dump.log -D ./ -t 494959
2021-05-30T13:36:43.8314025Z + setsid ./loop.sh
```
To solve this problem, we could use command substitution to preserve the order of execution.
Bash performs expansion of `$( (command) )` in a subshell environment and replacing it with the standard output of the command. Hence, `PID` would be assigned after `setsid ./loop.sh` has started.
```bash
PID=$( (setsid ./loop.sh < /dev/null &> /dev/null) & jobs -r -p)
```
```
2021-06-06T11:12:02.5890420Z ++ ++ setsid ./loop.sh
2021-06-06T11:12:02.5894022Z jobs -r -p
2021-06-06T11:12:02.5900376Z + PID=658293
```

Reply:
These commands were not executed in the correct order on Alpine and Fedora Rawhide, as show below.
```
2021-05-30T13:36:43.8309467Z + PID=494959
2021-05-30T13:36:43.8310637Z + /criu/criu/criu dump -v4 -o dump.log -D ./ -t 494959
2021-05-30T13:36:43.8314025Z + setsid ./loop.sh
```

I don't believe this is an order of execution problem, but rather some logging race.

The actual problem I think is `setsid`. Quoting `setsid(1)` man:

> The command calls fork(2) if already a process group leader.

So it might or might not fork, depending on the environment. I did a quick test and in my terminal it does not -- but we can add `-f` to make it fork.

```console
[kir@kir-rhat criu]$ cat x
setsid ls -ld /proc/self &
echo PID=$!
[kir@kir-rhat criu]$ bash -x x 
+ echo PID=986412
PID=986412
+ setsid ls -ld /proc/self
lrwxrwxrwx. 1 root root 0 Jun  1 14:02 /proc/self -> 986412
[kir@kir-rhat criu]$ vim x # Add -f to setsid so it always forks
[kir@kir-rhat criu]$ cat x
setsid -f ls -ld /proc/self &
echo PID=$!
[kir@kir-rhat criu]$ bash -x x
+ echo PID=986436
PID=986436
+ setsid -f ls -ld /proc/self
lrwxrwxrwx. 1 root root 0 Jun  1 14:02 /proc/self -> 986438
[kir@kir-rhat criu]$ 
```

As you can see, if `setsid -f` is used, PIDs don't match. Same happens in GHA environments even without `-f` I guess.

So this is the cause.

As to the solution, I think the best way is to have `loop.sh` write its PID to a file, and wait for this file in the parent (using a polling loop, or `inotifywait(1)`, or, better, use a pipe (as @avagin suggested above) to pass the PID. The implementation is left as an exercise for the reader; runc checkpoint.bats provides a good example.